### PR TITLE
Add Example: Simon's Algorithm

### PR DESCRIPTION
**Pull request by sebgrijalva**
This is an implementation of Simon's Algorithm using Cirq, including the quantum circuit and classical post-processing. Any feedback is highly appreciated!

### CODE REVIEW

**Comment by sebgrijalva** [comment]
@googlebot I signed it!

**Comment by vtomole** [comment]
@sebgrijalva Please fix the formatting errors by running `check/format-incremental --apply`

File: examples/simon_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+import random
+import cirq
+
+#needed for classical post-processing;
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Add a space between `#` and the comment

File: examples/simon_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import scipy as sp
+import numpy as np
+from collections import Counter
+
+"""Demonstrates the Simon algorithm.
```
Comment (vtomole) [inline examples/simon_algorithm.py]: `Demonstrates Simon's algorithm.`

File: examples/simon_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                   ││││           │   │
+(8, 0): ───────────X┼┼┼───────────┼───┼────────────────────
+                    │││           │   │
+(9, 0): ────────────X┼┼───────────X───×────────────────────
+                     ││
+(10, 0): ────────────X┼────────────────────────────────────
+                      │
+(11, 0): ─────────────X────────────────────────────────────
+                └──────┘   └───────────┘
+Most common Simon Algorithm answer is: ('[1 0 0 1 0 0]', 100)
```
Comment (vtomole) [inline examples/simon_algorithm.py]: `Most common Simon's Algorithm answer is`

File: examples/simon_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+***If the input string is s=0^n, no significant answer can be distinguished (since the
+null-space of the system of equations provided by the measurements gives a random vector). 
+In this case the output is: 
+"No significant answers obtained. Secret Sequence is probably [0,0,0,0,0,0]"
+"""
+
+
+def main(qubit_count=6):
+
+    data = [] #we'll store here the results
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Space between `#`.

File: examples/simon_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    #define a secret string:
+    secret_string = [random.randint(0, 1) for _ in range(qubit_count)]
+
+    print(f'Secret string = {secret_string}')
+
+    n_samples=100
+    for _ in range(n_samples):
+        flag = False #check if we have a linearly independent set of measures
+        while flag == False:
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Replace with `while not Flag`

File: examples/simon_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+def make_oracle(input_qubits, output_qubits, secret_string):
+    """Gates implementing the function f(a) = f(b) iff a+b=s""" 
+
+    #copy contents to output qubits:
+    for control_qubit, target_qubit in zip(input_qubits, output_qubits):
+        yield cirq.CNOT(control_qubit, target_qubit)
+
+    #create mapping:
+    if sum(secret_string) != 0:  #check if the secret string is non-zero
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Replace with `if not sum`

**Comment by vtomole** [comment]
Call this in [examples_test.py](https://github.com/quantumlib/Cirq/blob/master/examples/examples_test.py) to cover it.

**Comment by sebgrijalva** [comment]
@vtomole  Thanks a lot for the patience! I'm familiarizing with github's system

File: examples/simon_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+Most common Simon Algorithm answer is: ('[1 0 0 1 0 0]', 100)
+
+***If the input string is s=0^n, no significant answer can be distinguished (since the
+null-space of the system of equations provided by the measurements gives a random vector). 
+In this case the output is: 
+"No significant answers obtained. Secret Sequence is probably [0,0,0,0,0,0]"
+"""
+
+
+def main(qubit_count=6):
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Reducing the qubit_count to 3 will still get the point across while saving some time on our tests

File: examples/simon_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+    data = []  # we'll store here the results
+
+    #define a secret string:
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Put space between the `#`.

File: examples/simon_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    secret_string = [random.randint(0, 1) for _ in range(qubit_count)]
+
+    print(f'Secret string = {secret_string}')
+
+    n_samples = 100
+    for _ in range(n_samples):
+        flag = False  #check if we have a linearly independent set of measures
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Put space between the `#`.

File: examples/simon_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        while not flag:
+            # Choose qubits to use.
+            input_qubits = [cirq.GridQubit(i, 0) for i in range(qubit_count)
+                           ]  #input x
+            output_qubits = [
+                cirq.GridQubit(i + qubit_count, 0) for i in range(qubit_count)
+            ]  #output f(x)
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Put space between the `#`.

File: examples/simon_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    print(f'Secret string = {secret_string}')
+
+    n_samples = 100
+    for _ in range(n_samples):
+        flag = False  #check if we have a linearly independent set of measures
+        while not flag:
+            # Choose qubits to use.
+            input_qubits = [cirq.GridQubit(i, 0) for i in range(qubit_count)
+                           ]  #input x
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Put space between the `#`.

File: examples/simon_algorithm.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                solution = np.around(null_space, 3)  # chop very small values
+                minval = abs(min(solution[np.nonzero(solution)], key=abs))
+                solution = (solution / minval % 2).astype(
+                    int)  # renormalize vector mod 2
+                data.append(str(solution))
+    freqs = Counter(data)
+    print('Circuit:')
+    print(circuit)
+    if freqs.most_common(1)[0][
+            1] < 0.15 * n_samples:  #forcing at least 15% of samples to coincide
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Put space between the `#`.

File: examples/simon_algorithm.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+                yield cirq.CNOT(input_qubits[significant], output_qubits[j])
+    # Apply a random permutation:
+    pos = [
+        0, 3
+    ]  # Choose any combination of qubits to swap to define the oracle. Here we choose 0 and 3:
+    yield cirq.SWAP(output_qubits[pos[0]], output_qubits[pos[1]])
+
+
+def make_simon_circuit(input_qubits, output_qubits, oracle):
+    """Solves for secret period s of a 2-to-1 function such that f(x)=f(y) iff x+y=s"""
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Use modulo addition symbol instead of `+`

File: examples/simon_algorithm.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            f'No significant answers obtained. Secret Sequence is probably {np.zeros(qubit_count)}'
+        )
+    else:
+        print(
+            f'Most common Simon{chr(39)}s Algorithm answer is: {freqs.most_common(1)[0]}'
+        )
+
+
+def make_oracle(input_qubits, output_qubits, secret_string):
+    """Gates implementing the function f(a) = f(b) iff a+b=s"""
```
Comment (vtomole) [inline examples/simon_algorithm.py]: Use modulo addition symbol instead of `+`.

**Comment by vtomole** [comment]
@sebgrijalva Let's also get the build to pass.

**Comment by sebgrijalva** [comment]
I'm not so sure how to fix the failing checks ('Build Protos', 'Coverage Check') ...

**Comment by vtomole** [comment]
@sebgrijalva  Write an example that causes 

```python
    if freqs.most_common(
            1
    )[0][1] < 0.25 * n_samples
```

to be executed.

**Pull request merged**